### ISSUE TITLE
[Bug] "&"- and "|"-chained beartype validators not short-circuited

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#125]
While brutally pounding on issue https://github.com/beartype/beartype/issues/123 kindly submitted by typing brain-child @braniii, I couldn't help but notice this horrifyingly unexpected exception:

```python
# In "Bigger_Bigger_braniii.py":
import numpy as np
from beartype import beartype
from beartype.vale import Is
from numpy.typing import NDArray
from typing import Annotated

IsArray2D = Is[lambda arr: arr.ndim == 2]
IsArrayMatrix = Is[lambda arr: arr.shape[0] == arr.shape[1]]
IsArraySymmetric = Is[lambda arr: np.allclose(arr, arr.T)]

FloatArray = NDArray[np.floating]
Float2DArray         = Annotated[FloatArray, IsArray2D]
FloatMatrix          = Annotated[FloatArray, IsArray2D & IsArrayMatrix]
SymmetricFloatMatrix = Annotated[FloatArray, IsArray2D & IsArrayMatrix & IsArraySymmetric]

@beartype
def func(arr: SymmetricFloatMatrix) -> SymmetricFloatMatrix:
    return arr

mat = np.zeros((4,))
func(mat)
```

```bash
$ python3.10 Bigger_Bigger_braniii.py 
Traceback (most recent call last):
  File "/home/leycec/tmp/Bigger_Bigger_braniii.py", line 24, in <module>
    func(mat)
  File "<@beartype(__main__.func) at 0x7f948d70bb50>", line 47, in func
  File "/usr/lib/python3.10/site-packages/beartype/_decor/_error/errormain.py", line 292, in raise_pep_call_exception
    ).get_cause_or_none()
  File "/usr/lib/python3.10/site-packages/beartype/_decor/_error/_errorsleuth.py", line 344, in get_cause_or_none
    return get_cause_or_none(self)
  File "/usr/lib/python3.10/site-packages/beartype/_decor/_error/_pep/_errorpep593.py", line 87, in get_cause_or_none_annotated
    hint_diagnosis = hint_validator.get_diagnosis(
  File "/usr/lib/python3.10/site-packages/beartype/vale/_core/_valecorebinary.py", line 152, in get_diagnosis
    line_inner_operand_1 = self._validator_operand_1.get_diagnosis(
  File "/usr/lib/python3.10/site-packages/beartype/vale/_core/_valecorebinary.py", line 160, in get_diagnosis
    line_inner_operand_2 = self._validator_operand_2.get_diagnosis(
  File "/usr/lib/python3.10/site-packages/beartype/vale/_core/_valecore.py", line 429, in get_diagnosis
    is_obj_valid=self.is_valid(obj),
  File "/home/leycec/tmp/Bigger_Bigger_braniii.py", line 10, in <lambda>
    IsArrayMatrix = Is[lambda arr: arr.shape[0] == arr.shape[1]]
IndexError: tuple index out of range
```

:facepalm:

# What Now? Let Me Tell You What Now.

To resolve this, we'll need to improve our internal error handler to either:

1. Catch and silently ignore *all* exceptions raised by validators. Although usually dangerous (thus, bad), doing so is guaranteed to be safe in this edge case. Why? Because any validator raising an unexpected exception would have already done so during earlier type-checking. It follows that any exceptions raised after type-checking are safely ignorable.
1. Short-circuit boolean operations by explicitly *not* calling subsequent `&`-chained validators whose preceding validators returned `False`. This is considerably harder. Ain't nobody got volunteer time to get all De Morgan's Laws up in here.

Option **1,** I choose you!

**Comment by leycec** [comment]
Actually... Option **2.** is the obvious way. Short-circuiting is *always* the Right Thing™ to do. So, let's just do that thing. :+1:

**Comment by leycec** [comment]
**Brutally resolved by 809e43a.** Chained validators now short-circuit as expected, so:

```python
# This overly verbose validator...
IsArrayMatrix = Is[lambda arr: arr.ndim == 2 and arr.shape[0] == arr.shape[1]]

# ...can now be reduced to just this!
IsArrayMatrix = Is[lambda arr: arr.shape[0] == arr.shape[1]]
```

And that's how the West was won, @braniii. :fist_oncoming:
